examples/tcp-wait/server.rs

```diff
@@ -0,0 +1,239 @@
+
+#[cfg(target_os = "linux")]
+pub const SOCK_STREAM: i32 = libc::SOCK_STREAM;
+
+pub struct TcpServer {
+    /// Underlying libOS.
+    libos: LibOS,
+    /// Local socket descriptor.
+    sockqd: Option<QDesc>,
+    /// Number of clients expected to connect.
+    nclients: usize,
+    /// Connected clients.
+    clients: HashSet<QDesc>,
+    /// Pending operations.
+    qts: Vec<QToken>,
+    /// Reverse mapping of pending operations.
+    qts_reverse: HashMap<QToken, QDesc>,
+    /// Number of accepted connections.
+    clients_accepted: usize,
+    /// Number of closed connections.
+    clients_closed: usize,
+}
+
+impl TcpServer {
+    pub fn new(mut libos: LibOS, local: SocketAddrV4, nclients: usize) -> Result<Self> {
+        // Create TCP socket.
+        let sockqd: QDesc = libos.socket(AF_INET, SOCK_STREAM, 0)?;
+
+        // Bind to local address.
+        libos.bind(sockqd, local)?;
+
+        println!("Listening to: {:?}", local);
+
+        return Ok(Self {
+            libos,
+            sockqd: Some(sockqd),
+            nclients,
+            clients: HashSet::default(),
+            qts: Vec::default(),
+            qts_reverse: HashMap::default(),
+            clients_accepted: 0,
+            clients_closed: 0,
+        });
+    }
+
+    fn register_client(&mut self, qd: QDesc) {
+        assert_eq!(
+            self.clients.insert(qd),
+            true,
+            "client is already registered and it shouldn't be"
+        );
+    }
+
+    fn unregister_client(&mut self, qd: QDesc) {
+        assert_eq!(
+            self.clients.remove(&qd),
+            true,
+            "client isn't registered and it should be"
+        );
+    }
+
+    fn cancel_pending_operations(&mut self, qd: QDesc) -> Vec<QToken> {
+        let qts_drained: HashMap<QToken, QDesc> = self.qts_reverse.drain_filter(|_k, v| *v == qd).collect();
+        let qts_dropped: Vec<QToken> = self.qts.drain_filter(|x| qts_drained.contains_key(x)).collect();
+        qts_dropped
+    }
+
+    fn mark_completed_operation(&mut self, index: usize) -> Result<()> {
+        let qt: QToken = self.qts.remove(index);
+        self.qts_reverse
+            .remove(&qt)
+            .ok_or(anyhow::anyhow!("unregistered queue token"))?;
+        Ok(())
+    }
+
+    fn issue_accept(&mut self) -> Result<()> {
+        let qt: QToken = self.libos.accept(self.sockqd.expect("should be a valid socket"))?;
+        self.qts_reverse
+            .insert(qt, self.sockqd.expect("should be a valid socket"));
+        self.qts.push(qt);
+        Ok(())
+    }
+
+    fn issue_pop(&mut self, qd: QDesc) -> Result<()> {
+        let qt: QToken = self.libos.pop(qd, None)?;
+        self.qts_reverse.insert(qt, qd);
+        self.qts.push(qt);
+        Ok(())
+    }
+
+    fn issue_close(&mut self, qd: QDesc) -> Result<()> {
+        match self.libos.close(qd) {
+            Ok(_) => Ok(()),
+            Err(e) if e.errno == libc::ECONNRESET => Ok(()),
```

Comment:
You should remove this assert. When closing it is not possible to fail with `ECONNRESET`.